[mkdocs] docs: add complete aggregate tutorial (#1772)
symbol/symbol: mkdocs/pages/en/devbook/transactions/complete-aggregate.md

---
title: Complete Aggregate
---

# Creating a Complete Aggregate Transaction

This tutorial shows how to create an asset swap using <complete aggregate transactions:>.

In this example, Account A sends 10 <XYM:> to Account B, while Account B simultaneously sends 1 custom <mosaic:> back
to Account A:

```dot
digraph {
    rankdir="LR";
    fontsize=12;
    subgraph clusterAggregate {
        label = "Complete Aggregate Transaction";
        tooltip = "Complete Aggregate Transaction";
        subgraph clusterT2 {
            label = "Embedded Transfer 2";
            tooltip = "Embedded Transfer 2";
            style = dashed;
            A1 [label="Account A" tooltip="Account A"];
            B1 [label="Account B" tooltip="Account B"];
            A1 -> B1 [label="1 Custom Mosaic" dir=back];
        }
        subgraph clusterT1 {
            label = "Embedded Transfer 1";
            tooltip = "Embedded Transfer 1";
            style = dashed;
            A2 [label="Account A" tooltip="Account A"];
            B2 [label="Account B" tooltip="Account B"];
            A2 -> B2 [label="10 XYM"];
        }

> 

Reviewer: Put clusterT1 before clusterT2 so they appear in their natural order.
Remove empty line.

Reviewer: Weird. It doesn't happen in the public docs.
https://docs.symboltest.net/en/textbook/transactions/#aggregate-transactions

Different version of some tool?

Author: Edited, but it renders in the opposite order:

<img width="534" height="298" alt="image" src="https://github.com/user-attachments/assets/fe22c3fd-c067-4ea8-a16f-920cca23e8ba" />

It also happens in other pages:

<img width="719" height="544" alt="image" src="https://github.com/user-attachments/assets/357db493-821a-4872-8453-eb179118b36e" />

Author: Ok, should be an issue on my side.

    }
}
```

Both parties coordinate off-chain to collect signatures before announcement, ensuring the swap executes as a

> single atomic transaction.
> 
> 

Reviewer: ```suggestion

```
No multiple blank lines. Are you using the markdownlint extension?

!!! tip "When to use complete vs bonded aggregates"

    Complete aggregates work well in the following cases:

    - A single account batches multiple transactions.
    - Multiple parties coordinate off-chain to collect <signatures:> before publishing the transaction.

    If off-chain coordination is impractical, use <Bonded Aggregate Transactions:> instead,

>     which allow <cosignatures:> to be added on-chain.

Reviewer: Good idea, this is a useful info box.
Just a couple of things:
* "Bonded" appears for the first time in the title, so we cannot provide a link to explain what it is. I would remove it from the title.
* You're mixing what agg txes are good for (batching txes) with what complete agg txes are good for (coordinating offchain). This is a confusing topic, so we need to be careful.
* This is the ideal place to provide a concrete example that helps understand the tutorial.

So I suggest something like this. Feel free to adapt.

```suggestion
!!! tip "Two types of aggregate transactions"

    <Aggregate transactions:> group multiple <transactions:> in single operation, and require <signatures:> from all involved accounts.

    A _complete aggregate transaction_ collects all signatures before being announced,
    and therefore works best when parties can easily communicate off-chain and share the same signing tools.
    For example, inside the same company.

    If off-chain coordination is impractical, use <Bonded Aggregate Transactions:> instead,
    which allow <cosignatures:> to be added on-chain.
```

Author: Edited part (title, explain what an aggregate is), but did not apply:  

- “share the same signing tools”:  Could be misleading because parties don't actually need to use the same tools. Each party signs with their own private key on their own machine using whatever SDK, wallet, or signing infrastructure they prefer.
- “for example, inside the same company”:  could be too prescriptive and potentially limiting. Could be done between comapnies / individuals if they can colaborate off-chain (e.g. via  a platform that makes that easy).  

Also, kept the use case “A single account batches multiple transactions”.

## Prerequisites

Before you start, make sure to:

> - Set up your development environment.

Reviewer: Use asterisks for unordered lists.

  See [Setting Up a Development Environment](../start/setup.md).

> - Create an <account:> to initiate the aggregate transaction, either

Reviewer: You need to create two accounts for this tutorial.

multiplier from <get:/network/fees/transaction>, following the same steps described in the
[Transfer Transaction](./transfer.md) tutorial.

### Creating Embedded Transactions

{{ tutorial.code_snippet(['py:58:79', 'js:56:79']) }}

The <embedded transactions:> define the operations to execute atomically.
Each embedded transaction specifies:

* **Type:** All transaction types can be embedded within aggregates (except other aggregates).
  For embedded transfers, use `transfer_transaction_v1`, the same as for basic transfer transactions.

* **Signer public key:** The account that would sign this transaction if it were announced
  independently.

* **Transaction-specific fields:** All fields specific to the transaction type must be provided.
  For transfers, this includes the recipient address and the <mosaics:> to send.

Note that embedded transactions do **not** include fee or deadline fields.
These are inherited from the enclosing aggregate transaction.

The example creates two <transfer transactions:> for the swap:

* The first transfer sends 10 <XYM:> from Account A to Account B.

> * The second transfer sends 1 custom <mosaic:> from Account B to Account A.

Reviewer: No need to link to mosaic again so soon.

### Building the Aggregate Transaction

{{ tutorial.code_snippet(['py:81:96', 'js:81:98']) }}

Once the embedded transactions are prepared, create the complete aggregate transaction that wraps them:

* **Type:** Use `aggregate_complete_transaction_v3`.

* **Signer public key:** The account initiating the aggregate.

>   This account announces the transaction and pays the transaction fee.

Reviewer: There's an interesting detail worth mentioning here, maybe even in an info box.
If the accounts want to share the cost of the transaction, B can include a XYM transfer to A to help pay the fee.
This can be done as an additional transaction or including it in the list of mosaics if the B->A embedded tx.

At some point we'll want a full tutorial about this technique, because it also enables sending transactions for free by having another account pay the fees.

> * **Deadline:** The maximum time the network should attempt to confirm the transaction.

Reviewer: ```suggestion
* **Deadline:** The timestamp, in network time, after which the transaction expires and can no longer be confirmed.
```

And link `network time` to the place where it is defined in the Transfer Transaction tutorial.

* **Transactions hash:** A hash computed from all embedded transactions.
  This ensures the embedded transactions cannot be modified after signing.
  Use <dy:SymbolFacade.hashEmbeddedTransactions> to compute this value.

* **Transactions:** The array of embedded transactions to execute.

> The <fee:> is calculated based on the aggregate's total size, which includes all embedded transactions plus

Reviewer: Oops! We don't have a definition for `fee` yet!
I'll add one, and add links to it everywhere, but meanwhile, we better remove this link.

space reserved for <cosignatures:> (104 bytes each).

### Collecting Signatures

{{ tutorial.code_snippet(['py:98:134', 'js:100:140']) }}

With the aggregate built, both accounts must sign it off-chain before announcement.

The snippet shows this process in distinct sections showing what happens on each machine:

1. **Account A (Initiator)** signs the transaction using <dy:SymbolFacade.signTransaction> and attaches the
signature using <dy:SymbolTransactionFactory.attachSignature> to create a shareable payload.
Account A then sends this payload to Account B through an off-chain channel.

2. **Account B (Cosignatory)** receives the payload and deserializes it using
<dy:SymbolTransactionFactory.deserialize> to reconstruct the transaction object.
Account B then cosigns using <dy:SymbolFacade.cosignTransaction>, which computes the transaction hash and
signs it.
The cosignature is sent back to Account A.

3. **Account A** receives the cosignature and adds it to the transaction object's cosignatures array,

> then rebuilds the payload.

Reviewer: ```suggestion
With the aggregate transaction built, both accounts must sign it off-chain before it can be announced.
The snippet above separates the process by machine:

1. **Account A (Initiator)** signs the transaction using <dy:SymbolFacade.signTransaction>.
It then uses <dy:SymbolTransactionFactory.attachSignature> which normally produces a fully announce-ready payload.
In this case, however, the payload is still missing Account B’s cosignature.
Account A sends this intermediate payload to Account B through an off-chain channel.

2. **Account B (Cosignatory)** receives the payload and deserializes it using
<dy:SymbolTransactionFactory.deserialize> to reconstruct the transaction object.
Account B should verify that the embedded transactions match what it expects to sign.
It then cosigns using <dy:SymbolFacade.cosignTransaction>, which computes the transaction hash and
produces a cosignature object.
Only this cosignature is sent back to Account A.

3. **Account A** receives the Account B's cosignature, adds it to the transaction object's `cosignatures` array,
and rebuilds the payload for announcement.
```

I felt this section was a bit convoluted.

> !!! note "Single-signer complete aggregates"

Reviewer: Again, good point! This is a useful bit of info.

We can complete it by explaining that if an account needs to sign multiple embedded transactions, only one signature is needed.
For example, Account A only needs to sign once, even though its signature is needed for the aggregate tx AND the first embedded tx.
And link to https://docs.symboltest.net/en/textbook/transactions/#embedded-transactions
And the title of the box needs to be changed then.

    When all embedded transactions share the same signer, such as batching multiple operations
    from one account, cosignatures are **not** required.

    In this case, the complete aggregate transaction can be announced immediately after signing, and the fee
    calculation does not need to reserve space for cosignatures.

### Announcing the Transaction

> 

Reviewer: Before the snippet, add that: now that the transaction is ready to be announced, it follows the same process as regular, non-aggregate transactions, as shown in the Transfer Transaction tutorial.

{{ tutorial.code_snippet(['py:136:149', 'js:142:154']) }}

Once all signatures are collected, the transaction is announced to a <node:> using the
<put:/transactions> endpoint.

The node validates that all required signatures are present and valid before accepting the transaction.
If validation passes, the transaction is added to the <unconfirmed pool:> and broadcast to other nodes.

### Waiting for Confirmation

{{ tutorial.code_snippet(['py:151:171', 'js:156:180']) }}

After announcement, the transaction status is monitored using <get:/transactionStatus/{hash}>.

The polling loop checks the status every second until the transaction is confirmed or fails.
Once confirmed, the swap is complete and both transfers have executed.

> 

Reviewer: Add that, for more information about announcing and awaiting confirmation the reader should check the Transfer Transaction tutorial.

## Output

The output shown below corresponds to a typical run of the program.

```text

> --8<-- 'devbook/transactions/complete-aggregate.log'

Reviewer: This is quite a long output log. It will be more useful if you add line numbers and highlight some lines, and then comment on those lines.

```md
   ```text linenums="1" hl_lines="10 20 30"
```

For example, I think these lines are worth explaining:
* `"type"` indicates that this is an aggregate.
* `"transactions"` contains the two embedded txes.
* `"signature": "0000"` contains the signature for the aggregate.
* `"cosignatures": []` contains any other cosignatures. If you output the transaction once the signatures have been added instead of at the beginning, these fields will make more sense, and you can add "note how Account A's signature is only needed once".
* `Waiting for confirmation` contains the hash of the transaction, in case readers wants to check it in the explorer.

And any others you see fit.

```

The aggregate transaction is treated as a single atomic unit by the network.
The swap executes completely: Account A receives the custom mosaic and Account B receives the XYM,
or the entire transaction fails and no assets are transferred.

You can view the transactions on the [Symbol Testnet Explorer](https://testnet.symbol.fyi/) by searching for the
aggregate transaction hash announced.

## Conclusion

This tutorial showed how to:

| Step                                                                   | Related documentation                                                               |
| ---------------------------------------------------------------------- | ------------------------------------------------------------------------------------|

> | [Obtain deadline and fee information](#fetching-network-time-and-fees) | <get:/node/time><br/><get:/network/fees/transaction>                                |

Reviewer: But this was not the purpose of this tutorial, so I wouldn't list this.

| [Create embedded transactions](#creating-embedded-transactions)        | <dy:SymbolTransactionFactory.createEmbedded>                                        |
| [Build the aggregate](#building-the-aggregate-transaction)             | <dy:SymbolTransactionFactory.create><br/><dy:SymbolFacade.hashEmbeddedTransactions> |
| [Collect signatures off-chain](#collecting-signatures)                 | <dy:SymbolFacade.signTransaction><br/><dy:SymbolFacade.cosignTransaction>           |
| [Announce the transaction](#announcing-the-transaction)                | <put:/transactions>                                                                 |

> | [Wait for confirmation](#waiting-for-confirmation)                     | <get:/transactionStatus/{hash}>                                                     |

Reviewer: Same with these two, I would remove them.
People reading this tutorial should already know how to do this. And if they don't, we've given them plenty opportunities to read the more basic tutorial.

In this example, Account A sends 10 <XYM:> to Account B, while Account B simultaneously sends 1 custom <mosaic:> back
to Account A:

```dot
digraph {
    rankdir="LR";
    fontsize=12;
    subgraph clusterAggregate {
        label = "Complete Aggregate Transaction";
        tooltip = "Complete Aggregate Transaction";
        subgraph clusterT1 {
            label = "Embedded Transfer 1";
            tooltip = "Embedded Transfer 1";
            style = dashed;
            A2 [label="Account A" tooltip="Account A"];
            B2 [label="Account B" tooltip="Account B"];
            A2 -> B2 [label="10 XYM"];
        }
        subgraph clusterT2 {
            label = "Embedded Transfer 2";
            tooltip = "Embedded Transfer 2";
            style = dashed;
            A1 [label="Account A" tooltip="Account A"];
            B1 [label="Account B" tooltip="Account B"];
            A1 -> B1 [label="1 Custom Mosaic" dir=back];
        }
    }
}
```

Both parties coordinate off-chain to collect signatures before announcement, ensuring the swap executes as a
single atomic transaction.

!!! note "Two types of aggregate transactions"

>     <Aggregate transactions:> group multiple <transactions:> in single operation, and require <signatures:> from all

Reviewer: ```suggestion
    <Aggregate transactions:> group multiple <transactions:> in a single operation, and require <signatures:> from all
```

    involved accounts.

    A _complete aggregate transaction_ collects all signatures before being announced.

>     This works well in two scenarios:

Reviewer: ```suggestion
    This procedure works well, for example, in the following scenarios:
```

I don't want to suggest that these are the _only_ scenarios where this works well.
Also, "naked this" are to be avoided.

    * **Multi-party coordination:** When parties can communicate off-chain to exchange transaction payloads and

>       cosignatures.

Reviewer: This is now the first usage of the term "cosignature", so the link to the definition should be moved here.

Also, I realize the definition of cosignature is a bit ambiguous: it should clarify that it applies to multiple signatures from different accounts. I'll fix that in the Textbook.

    * **Single-account batching:** When one account wants to execute multiple transactions atomically.
      No cosignatures are needed in this case.

    If off-chain coordination is impractical, use <bonded aggregate transactions:> instead,
    which allow <cosignatures:> to be added on-chain.

## Prerequisites

Before you start, make sure to:

* Set up your development environment.
  See [Setting Up a Development Environment](../start/setup.md).
* Create an <account:> (Account A) to initiate the aggregate transaction, either
  [from code](../accounts/create-from-private-key.md) or
  [by using a wallet](../../userbook/wallet/create-account.md).
* Create a second account (Account B) to participate in the swap.
* Obtain <XYM:> for Account A to pay for the transaction fee and transfer amounts.
  See [Getting Testnet Funds from the Faucet](../accounts/testnet-faucet.md).

> * Create a <mosaic:> owned by Account B for the swap.

Reviewer: This long list of prerequisites looks a bit scary. Maybe we can split it like this:

```suggestion
* Set up your development environment.
  See [Setting Up a Development Environment](../start/setup.md).
  
This tutorial requires two accounts with adequate funds to perform the transfers.
Although pre-funded accounts are provided for convenience, they are not maintained and may run out of funds.
If so, you will need to:

* Create an <account:> (Account A) to initiate the aggregate transaction, either
  [from code](../accounts/create-from-private-key.md) or
  [by using a wallet](../../userbook/wallet/create-account.md).
* Create a second account (Account B) to participate in the swap.
* Obtain <XYM:> for Account A to pay for the transaction fee and transfer amounts.
  See [Getting Testnet Funds from the Faucet](../accounts/testnet-faucet.md).
* Create a <mosaic:> owned by Account B for the swap.
```

2. **Account B (Cosignatory)** receives the payload and deserializes it using
  <dy:SymbolTransactionFactory.deserialize> to reconstruct the transaction object.
  Account B should verify that the embedded transactions match what it expects to sign.
  It then cosigns using <dy:SymbolFacade.cosignTransaction>, which computes the transaction hash and produces a
  cosignature object.
  Only this cosignature is sent back to Account A.

3. **Account A** receives the Account B's cosignature, adds it to the transaction object's `cosignatures` array, and
  rebuilds the payload for announcement.

!!! note "Signatures in aggregate transactions"

    An account only signs once, even if it appears as the signer in multiple embedded transactions.
    In this tutorial, Account A signs the aggregate transaction, which covers both the aggregate itself and the
    first embedded transaction where Account A is the signer.

    When all embedded transactions share the same signer (batching multiple operations from one account),
    cosignatures are **not** required. The aggregate can be announced immediately after signing, and the fee
    calculation does not need to reserve space for cosignatures.

>     See [Embedded Transactions](../../textbook/transactions.md#embedded-transactions) for more details.

Reviewer: Better use a definition `<embedded transactions:>`. It's shorter, path-independent, and provides a nice tooltip (although at this point in the tutorial it shouldn't be necessary).

### Announcing the Transaction

Now that the transaction is ready to be announced, it follows the same process as regular, non-aggregate transactions,
as shown in the [Transfer Transaction](./transfer.md#announcing-the-transaction) tutorial.

{{ tutorial.code_snippet(['py:135:148', 'js:140:152']) }}

Once all signatures are collected, the transaction is announced to a <node:> using the
<put:/transactions> endpoint.

The node validates that all required signatures are present and valid before accepting the transaction.
If validation passes, the transaction is added to the <unconfirmed pool:> and broadcast to other nodes.

### Waiting for Confirmation

{{ tutorial.code_snippet(['py:150:170', 'js:154:178']) }}

After announcement, the transaction status is monitored using <get:/transactionStatus/{hash}>.

The polling loop checks the status every second until the transaction is confirmed or fails.
Once confirmed, the swap is complete and both transfers have executed.

## Output

The output shown below corresponds to a typical run of the program.

```text linenums="1" hl_lines="10 14 18 48 53 60 66"
--8<-- 'devbook/transactions/complete-aggregate.log'
```

> Key points in the output:

Reviewer: The idea behind adding line numbers to the output log was to reference them now :)

For example:
```md
* Line 10 (`"signature": "0000..."`): Shows all zeros...
```
Or:
```md
* Line 10, `"signature": "0000..."`
    Shows all zeros...
```

* `"signature": "0000..."`: Shows all zeros initially because the transaction hasn't been signed yet.
* `"type": 16705`: Indicates this is an `aggregate_complete_transaction_v3`.
* `"transactions"`: Contains the two embedded transfers that will execute atomically.
* `"cosignatures": []`: Initially empty. Account B's cosignature is added before announcement.

>   Note how Account A's signature is only needed once, even though it appears as signer in both the aggregate and the 

Author: ```suggestion
  Note how Account A's signature is only needed once, even though it appears as signer in both the aggregate and the
```

Extra whitespace, will fix after the review.

  first embedded transaction.

> * `"payload": "8010..."`: The transaction payload computed from the aggregate transaction and its embedded transactions.

Reviewer: ```suggestion
* `"payload": "68010..."`: The transaction payload computed from the aggregate transaction and its embedded transactions.
```
